core/src/smartcontracts/isi/permissions/judge.rs

```diff
@@ -0,0 +1,459 @@
+//! Module with [`Judge`] trait and its implementations
+
+use std::sync::Arc;
+
+use super::*;
+
+/// Boxed generic judge
+pub type OperationJudgeBoxed<O> = Box<dyn Judge<Operation = O> + Send + Sync>;
+/// Boxed [`Instruction`] judge
+pub type InstructionJudgeBoxed = OperationJudgeBoxed<Instruction>;
+/// Boxed [`QueryBox`] judge
+pub type QueryJudgeBoxed = OperationJudgeBoxed<QueryBox>;
+/// Boxed [`Expression`] judge
+pub type ExpressionJudgeBoxed = OperationJudgeBoxed<Expression>;
+
+/// [`Arc`] with generic judge
+pub type OperationJudgeArc<O> = Arc<dyn Judge<Operation = O> + Send + Sync>;
+/// [`Arc`] with [`Instruction`] judge
+pub type InstructionJudgeArc = OperationJudgeArc<Instruction>;
+/// [`Arc`] with [`QueryBox`] judge
+pub type QueryJudgeArc = OperationJudgeArc<QueryBox>;
+/// [`Arc`] with [`Expression`] judge
+pub type ExpressionJudgeArc = OperationJudgeArc<Expression>;
+
+/// The judge that gives the final decision
+/// whether or not the `operation` should be accepted.
+///
+/// Unlike [`IsAllowed`], this trait requires the [`Result`] to be returned.
+///
+/// The judge accumulates [`verdicts`](ValidatorVerdict) from all validators,
+/// makes a decision, and returns the result.
+pub trait Judge: std::fmt::Debug {
```

Comment:
Debug is almost never a good supertrait. Instead consider a blanket impl if the underlying type `impl Debug`.

Reply:
Ok, it worked for `Judge`.
But it won't work for `IsAllowed` because we want to use `Debug` to construct an error message